native-engine/datafusion-ext-plans/src/agg/spark_udaf_wrapper.rs

```diff
@@ -0,0 +1,344 @@
+use crate::{
+    agg::{
+        acc::{AccColumn, AccColumnRef},
+        agg::{Agg, IdxSelection},
+    },
+    idx_for_zipped,
+    memmgr::spill::{SpillCompressedReader, SpillCompressedWriter},
+};
+
+pub struct SparkUDAFWrapper {
+    serialized: Vec<u8>,
+    pub return_type: DataType,
+    child: Vec<Arc<dyn PhysicalExpr>>,
+    import_schema: SchemaRef,
+    params_schema: OnceCell<SchemaRef>,
+    jcontext: OnceCell<GlobalRef>,
+}
+
+impl SparkUDAFWrapper {
+    pub fn try_new(
+        serialized: Vec<u8>,
+        return_type: DataType,
+        child: Vec<Arc<dyn PhysicalExpr>>,
+    ) -> Result<Self> {
+        Ok(Self {
+            serialized,
+            return_type: return_type.clone(),
+            child,
+            import_schema: Arc::new(Schema::new(vec![Field::new("", return_type, true)])),
+            params_schema: OnceCell::new(),
+            jcontext: OnceCell::new(),
+        })
+    }
+
+    fn jcontext(&self) -> Result<GlobalRef> {
+        self.jcontext
+            .get_or_try_init(|| {
+                let serialized_buf = jni_new_direct_byte_buffer!(&self.serialized)?;
+                let jcontext_local =
+                    jni_new_object!(SparkUDAFWrapperContext(serialized_buf.as_obj()))?;
+                jni_new_global_ref!(jcontext_local.as_obj())
+            })
+            .cloned()
+    }
+}
+
+impl Display for SparkUDAFWrapper {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "SparkUDAFWrapper")
+    }
+}
+
+impl Debug for SparkUDAFWrapper {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "SparkUDAFWrapper({:?})", self.child)
+    }
+}
+
+impl Agg for SparkUDAFWrapper {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn exprs(&self) -> Vec<Arc<dyn PhysicalExpr>> {
+        self.child.clone()
+    }
+
+    fn data_type(&self) -> &DataType {
+        &self.return_type
+    }
+
+    fn nullable(&self) -> bool {
+        true
+    }
+
+    fn create_acc_column(&self, num_rows: usize) -> AccColumnRef {
+        let jcontext = self.jcontext().unwrap();
+        let rows = jni_call!(SparkUDAFWrapperContext(jcontext.as_obj()).initialize(
+            num_rows as i32,
+        )-> JObject)
+        .unwrap();
+
+        let jcontext = self.jcontext().unwrap();
+        let obj = jni_new_global_ref!(rows.as_obj()).unwrap();
+        Box::new(AccUnsafeRowsColumn {
+            obj,
+            jcontext,
+            num_rows,
+        })
+    }
+
+    fn with_new_exprs(&self, _exprs: Vec<Arc<dyn PhysicalExpr>>) -> Result<Arc<dyn Agg>> {
+        Ok(Arc::new(Self::try_new(
+            self.serialized.clone(),
+            self.return_type.clone(),
+            self.child.clone(),
+        )?))
+    }
+
+    fn partial_update(
+        &self,
+        accs: &mut AccColumnRef,
+        acc_idx: IdxSelection<'_>,
+        partial_args: &[ArrayRef],
+        partial_arg_idx: IdxSelection<'_>,
+        batch_schema: SchemaRef,
```

Comment:
i suggest to create `params_schema` in constructor and remove this `batch_schema` parameter.